File: 킬링캠프/2주 차 - 리스트, 큐, 스택, 그래프, DFS, BFS/challenge/Coin Change/Solution.java
import java.util.*;

class Solution {
    public int coinChange(int[] coins, int amount) {
        if (amount == 0) {
            return 0;
        }

        return bfs(coins, amount);
    }

    private int bfs(int[] coins, int amount) {
        boolean[] visited = new boolean[amount + 1];
        Deque<Vertex> dq = new ArrayDeque<>();
        dq.offer(new Vertex(amount, 0));
        visited[amount] = true;

        while (!dq.isEmpty()) {
            Vertex currentVertex = dq.poll();

            for (int coin : coins) {
                int nextAmount = currentVertex.amount - coin;
                int nextCoinCount = currentVertex.coinCount + 1;
                if (nextAmount == 0) {
                    return nextCoinCount;
                }

                if (nextAmount > 0 && !visited[nextAmount]) {
                    dq.offer(new Vertex(nextAmount, nextCoinCount));
                    visited[nextAmount] = true;
                }
            }
        }

        return -1;
    }
}

class Vertex {
    int amount;
    int coinCount;

    Vertex(int amount, int coinCount) {
        this.amount = amount;
        this.coinCount = coinCount;
    }
}

File: 킬링캠프/2주 차 - 리스트, 큐, 스택, 그래프, DFS, BFS/challenge/Coin Change/solution.py
from collections import deque

class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        dq = deque()
        dq.append((amount, 0))
        visited = {amount}
        
        while dq:
            current_amount, count = dq.popleft()

            if current_amount == 0:
                return count
                
            if current_amount < 0:
                continue
                
            for coin in coins:
                next_amout = current_amount - coin
                if next_amout in visited:
                    continue

                dq.append((next_amout, count + 1))
                visited.add(next_amout)

        return -1
